lib/handler/range.c

```diff
@@ -0,0 +1,119 @@
+#include <assert.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "h2o.h"
+
+#if __GNUC__ >= 3
+# define likely(x) __builtin_expect(!!(x), 1)
+# define unlikely(x) __builtin_expect(!!(x), 0)
+#else
+# define likely(x) (x)
+# define unlikely(x) (x)
+#endif
+
+#define CHECK_EOF()                             \
+    if (buf == buf_end) {                       \
+        *ret = -2;                              \
+        return NULL;                            \
+    }
+
+#define EXPECT_CHAR(ch)                         \
+    CHECK_EOF();                                \
+    if (*buf++ != ch) {                         \
+        *ret = -1;                              \
+        return NULL;                            \
+    }
+
+size_t *process_range(h2o_mem_pool_t *pool, h2o_iovec_t *range_value, size_t file_size, size_t* ret)
+{
+    size_t range_start = -1, range_count = 0;
+    char *buf = range_value->base, *buf_end = buf + range_value->len;
+    int good_range=1;
+    H2O_VECTOR(size_t) ranges = {};
+    
+    EXPECT_CHAR('b');
+    EXPECT_CHAR('y');
+    EXPECT_CHAR('t');
+    EXPECT_CHAR('e');
+    EXPECT_CHAR('s');
+    EXPECT_CHAR('=');
+    /* most range requests contain only one range */
+    do {
+        range_start = -1; range_count = 0;
+        if (likely(*buf >= '0') && likely(*buf <= '9')) {
+            range_start = 0;
+            while (likely(*buf >= '0') && likely(*buf <= '9')) {
+                range_start *= 10;
+                range_start += *buf++ - '0';
+                CHECK_EOF();
+            }
```

Comment:
Please extract the logic of parsing a number to a dedicated function, and place it next to `h2o_strtosize` in lib/common/string.c.

I would be happy if the name and behavior of the function was modeled similar to `h2o_strtosize` with the extension that the end position of the string can be obtained.